### PR TITLE
Fix block drawing glitch at block boundaries

### PR DESCRIPTION
**Pull request by Rogier-5**
screenshot_20161218_151631

Sometimes, part of the surface of a mapblock is not rendered, while
all surrounding blocks are. This can be caused by the following:

1) Blocks that are at some distance from the player, and that consist
  of only air will not be sent to the client.
2) If the air/non-air transition is at the block's boundary, then the
  client will not render the boundary until it gets the air block too.
  Which it won't.

This patch fixes this, by using some of the leftover bandwidth to send
air-only blocks that are adjacent to non-air-only blocks. Thus, eventually,
the client will get the air-only blocks it requires, and draw the
block boundary of the adjacent block too.

### CODE REVIEW

**Comment by Fixer-007** [comment]
I knew this bug for very long time but was not sure if that was left out because of some constraints. I would like to see it fixed :+1: I guess I will also try it.

**Comment by Rogier-5** [comment]
Fixed C++11 compilation

File: src/clientiface.cpp [hunk 1]
```diff
@@ -335,6 +325,60 @@ void RemoteClient::GetNextBlocks (
...
+				Block is near ground level if night-time mesh
+				differs from day-time mesh.
+			*/
+			if (block && d >= d_opt && !block->getDayNightDiff()) {
+				// getDayNightDiff() does not take adjacent blocks into account.
+				// If necessary, remember this block, so that it may be sent anyway if possible.
+				if (extra_blockpos.count(p))
+					continue;
+				int extra = -1;
+				do {		// Can break ot of this
```
Comment (SmallJoker) [inline src/clientiface.cpp]: *out

Is this way preferred over `goto`?

Comment (Rogier-5) [inline src/clientiface.cpp]: IDK.
IMO there is more structure to this, and it is therefore easier to comprehend.

File: src/irr_v3d.h [hunk 2]
```diff
@@ -29,5 +29,24 @@ typedef core::vector3d<s16> v3s16;
 typedef core::vector3d<u16> v3u16;
 typedef core::vector3d<s32> v3s32;
 
+#if __cplusplus >= 201103L
+
+#include<functional>
```
Comment (Rogier-5) [inline src/irr_v3d.h]: fixed

**Comment by Rogier-5** [comment]
I have been thinking some more about this, and it seems to me the original code is basically a hack to begin with. It avoids sending air-blocks, which is sensible and sound (except for the bug that this PR tackles), but also underground blocks, IIUC by the assumption that more distant underground blocks will *necessarily* be obscured by other underground blocks closer by. That is an ugly hack. The original code fails in large caves: there being no day-night difference, many blocks will not be sent, and more distant parts of the cave will not be rendered, whereas they would be rendered if above ground.

Screenshots with and without the code in question:
screenshot_20161218_212143
screenshot_20161218_212231


Given that the original code is a hack, this fix is merely a workaround, that avoids the issue in *some* cases.

Any thoughts on removing the day-night check code altogether, and replacing it with just an air/no-air check ? Unfortunately, such a change would result in more underground blocks being sent to the client, often but certainly not always unnecessarily...

Maybe something can be done to optimize the underground case as well.

**Comment by paramat** [comment]
The day-night light difference check now has a setting so you can choose at what distance it applies, the default is 4 mapblocks. Surely occlusion culling should take care of this instead?

**Comment by paramat** [comment]
> Blocks that are at some distance from the player, and that consist
of only air will not be sent to the client.

Are you saying the day-nght diff check returns no diff for an all-air block? I would have thought air would have differing light values in daytime?

**Comment by Rogier-5** [comment]
> > Blocks that are at some distance from the player, and that consist of only air will not be sent to the client.
> 
> Are you saying the day-nght diff check returns no diff for an all-air block? I would have thought air would have differing light values in daytime?

Indeed. If the day-night check for any node in a block returns a difference, then an additional check is made whether the block contains only air. If it does, then the block's day-night check will return no difference, even though there *is* a difference...

The block's day-night check is really a check whether the block is exposed to sunlight *and* whether it contains anything that might be visible. So it's basically a heuristic to decide whether the block might be needed at all by the client when the player is at some distance, or whether we can get away with keeping it from the client in that case. It's not at all about whether there's a day-night difference... A better name would be 'mayBeNeededByTheClientWhenFarAway()` :-)

Occlusion culling can only take care of this as far as it can be performed on the server (maybe it can; on a block-level ?). The purpose of the code is apparently to save network traffic, not to save client-side CPU-time.

**Comment by paramat** [comment]
Ah oops i confused this with clientside occlusion culling.

**Comment by lhofhansl** [comment]
Do we still need #4686 with this?

**Comment by paramat** [comment]
I think so, as this sends air-only blocks, whereas #4686 allows partial-air and underwater blocks to be rendered.

**Comment by Fixer-007** [comment]
Hmmmm, I decided to test this on that water scene, and it seems there are much more stuff rendering with this PR.
with this PR:
with4925
without this PR:
without4925

**Comment by paramat** [comment]
@lhofhansl any further thoughts on this?

**Comment by lhofhansl** [comment]
Hmm... Not sure. With #5320, we could default the setting in #4686 to a larger value, which would have the same effect, with perhaps the same number of a extra blocks sent and rendered.

This is also related to #5239, which we said many neighboring blocks to get the lighting right.

If this sends significantly fewer blocks than increasing block_send_optimize_distance then it would be helpful, if not we should close this.

**Comment by paramat** [comment]
This is still very WIP and under consideration, not a priority for release.

**Comment by paramat** [comment]
@lhofhansl is this still needed, as your setting for block optimisation distance has been added?

**Comment by paramat** [comment]
Closing, neglected and Fixer's testing may show a problem.
@lhofhansl is the person to deal with this if necessary.

**Pull request closed without merging**